devito/types/object.py

```diff
@@ -220,3 +341,87 @@ def _mem_internal_eager(self):
     @property
     def _mem_internal_lazy(self):
         return self._liveness == 'lazy'
+
+
+class PetscObject(AbstractObjectWithShape):
+
+    is_PetscObject = True
+
+    def __init__(self, name, dtype, **kwargs):
+        self.name = name
+        self._dtype = dtype
+        self._is_const = kwargs.get('is_const', False)
+
+    # taken exactly from Function class but removed the "shape_global"
+    #  # and also added the option for all dimensions, shape and grid to be None
+    @classmethod
+    def __shape_setup__(cls, **kwargs):
+        grid = kwargs.get('grid')
+        dimensions = kwargs.get('dimensions')
+        shape = kwargs.get('shape')
+
+        if dimensions is None and shape is None and grid is None:
+            return None
+
+        if grid is None:
+            if shape is None:
+                raise TypeError("Need either `grid` or `shape`")
+        elif shape is None:
+            if dimensions is not None and dimensions != grid.dimensions:
+                raise TypeError("Need `shape` as not all `dimensions` are in `grid`")
+            shape = grid.shape_local
+        elif dimensions is None:
+            raise TypeError("`dimensions` required if both `grid` and "
+                            "`shape` are provided")
+        else:
+            # Got `grid`, `dimensions`, and `shape`. We sanity-check that the
+            # Dimensions in `dimensions` also appearing in `grid` have same size
+            # (given by `shape`) as that provided in `grid`
+            if len(shape) != len(dimensions):
+                raise ValueError("`shape` and `dimensions` must have the "
+                                 "same number of entries")
+            loc_shape = []
```

Comment:
no distinction between global and local shape (MPI irrelevant for Objects), so we don't need this part